File: app/src/test/java/io/github/hidroh/materialistic/LoginActivityTest.java
package io.github.hidroh.materialistic;

import android.content.Intent;
import android.support.design.widget.TextInputLayout;
import android.view.View;
import android.widget.EditText;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.shadows.ShadowAccountManager;
import org.robolectric.shadows.ShadowToast;
import org.robolectric.util.ActivityController;

import java.io.IOException;

import javax.inject.Inject;

import io.github.hidroh.materialistic.accounts.UserServices;
import io.github.hidroh.materialistic.test.TestRunner;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static org.assertj.android.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;

@RunWith(TestRunner.class)
public class LoginActivityTest {
    private ActivityController<LoginActivity> controller;
    private LoginActivity activity;
    @Inject UserServices userServices;
    @Captor ArgumentCaptor<UserServices.Callback> callback;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        TestApplication.applicationGraph.inject(this);
        reset(userServices);
        controller = Robolectric.buildActivity(LoginActivity.class);
        activity = controller.create().postCreate(null).start().resume().get();
    }

    @Test
    public void testEmptyLoginInput() {
        activity.findViewById(R.id.login_button).performClick();
        assertNotNull(((TextInputLayout) activity.findViewById(R.id.textinput_username)).getError());
        assertNotNull(((TextInputLayout) activity.findViewById(R.id.textinput_password)).getError());
    }

    @Test
    public void testEmptyRegisterInput() {
        activity.findViewById(R.id.register_button).performClick();
        assertNotNull(((TextInputLayout) activity.findViewById(R.id.textinput_username)).getError());
        assertNotNull(((TextInputLayout) activity.findViewById(R.id.textinput_password)).getError());
    }

    @Test
    public void testLoginSuccessful() {
        ((EditText) activity.findViewById(R.id.edittext_username)).setText("username");
        ((EditText) activity.findViewById(R.id.edittext_password)).setText("password");
        activity.findViewById(R.id.login_button).performClick();
        assertNull(((TextInputLayout) activity.findViewById(R.id.textinput_username)).getError());
        assertNull(((TextInputLayout) activity.findViewById(R.id.textinput_password)).getError());
        verify(userServices).login(eq("username"), eq("password"), eq(false), callback.capture());
        callback.getValue().onDone(true);
        assertThat(activity).isFinishing();
        assertEquals(activity.getString(R.string.welcome, "username"), ShadowToast.getTextOfLatestToast());
        assertThat(ShadowAccountManager.get(activity).getAccounts()).hasSize(1);
        assertEquals("username", Preferences.getUsername(activity));
    }

    @Test
    public void testRegisterFailed() {
        ((EditText) activity.findViewById(R.id.edittext_username)).setText("username");
        ((EditText) activity.findViewById(R.id.edittext_password)).setText("password");
        activity.findViewById(R.id.register_button).performClick();
        verify(userServices).login(eq("username"), eq("password"), eq(true), callback.capture());
        callback.getValue().onDone(false);
        assertThat(activity).isNotFinishing();
        assertEquals(activity.getString(R.string.login_failed), ShadowToast.getTextOfLatestToast());
    }

    @Test
    public void testLoginError() {
        ((EditText) activity.findViewById(R.id.edittext_username)).setText("username");
        ((EditText) activity.findViewById(R.id.edittext_password)).setText("password");
        activity.findViewById(R.id.login_button).performClick();
        verify(userServices).login(eq("username"), eq("password"), eq(false), callback.capture());
        callback.getValue().onError(new IOException());
        assertThat(activity).isNotFinishing();
        assertEquals(activity.getString(R.string.login_failed), ShadowToast.getTextOfLatestToast());
    }

    @Test
    public void testReLogin() {
        controller = Robolectric.buildActivity(LoginActivity.class);
        Preferences.setUsername(RuntimeEnvironment.application, "existing");
        activity = controller.create().postCreate(null).start().resume().get();
        assertThat(activity).hasTitle(R.string.re_enter_password);
        assertThat((View) activity.findViewById(R.id.register_button)).isNotVisible();
        assertThat((EditText) activity.findViewById(R.id.edittext_username))
                .hasTextString("existing");
    }

    @Test
    public void testAddAccount() {
        controller = Robolectric.buildActivity(LoginActivity.class);
        Preferences.setUsername(RuntimeEnvironment.application, "existing");
        Intent intent = new Intent();
        intent.putExtra(LoginActivity.EXTRA_ADD_ACCOUNT, true);
        activity = controller.withIntent(intent).create().postCreate(null).start().resume().get();
        assertThat(activity).hasTitle(R.string.title_activity_login);
        assertThat((View) activity.findViewById(R.id.register_button)).isVisible();
        assertThat((EditText) activity.findViewById(R.id.edittext_username)).isEmpty();
    }

    @After
    public void tearDown() {
        controller.pause().stop().destroy();
    }
}

File: app/src/test/java/io/github/hidroh/materialistic/LoginActivityTest.kt
package io.github.hidroh.materialistic

import android.accounts.Account
import android.accounts.AccountManager
import android.content.Intent
import com.google.android.material.textfield.TextInputLayout
import android.view.View
import android.widget.EditText
import io.github.hidroh.materialistic.accounts.UserServices
import org.assertj.android.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.eq
import org.mockito.Captor
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.android.controller.ActivityController
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowToast
import java.io.IOException

@RunWith(RobolectricTestRunner::class)
@Config(application = InjectableApplication::class)
class LoginActivityTest {
  private lateinit var controller: ActivityController<TestLoginActivity>
  private lateinit var activity: TestLoginActivity
  @Captor
  private lateinit var callback: ArgumentCaptor<UserServices.Callback>

  @Before
  fun setUp() {
    MockitoAnnotations.initMocks(this)
    controller = Robolectric.buildActivity(TestLoginActivity::class.java)
    activity = controller.create().get()
  }

  @Test
  fun testEmptyLoginInput() {
    activity.findViewById<View>(R.id.login_button).performClick()
    assertThat(activity.findViewById<TextInputLayout>(R.id.textinput_username).error).isNotNull()
    assertThat(activity.findViewById<TextInputLayout>(R.id.textinput_password).error).isNotNull()
  }

  @Test
  fun testEmptyRegisterInput() {
    activity.findViewById<View>(R.id.register_button).performClick()
    assertThat(activity.findViewById<TextInputLayout>(R.id.textinput_username).error).isNotNull()
    assertThat(activity.findViewById<TextInputLayout>(R.id.textinput_password).error).isNotNull()
  }

  @Test
  fun testLoginSuccessful() {
    activity.findViewById<EditText>(R.id.edittext_username).setText("username")
    activity.findViewById<EditText>(R.id.edittext_password).setText("password")
    activity.findViewById<View>(R.id.login_button).performClick()
    assertThat(activity.findViewById<TextInputLayout>(R.id.textinput_username).error).isNull()
    assertThat(activity.findViewById<TextInputLayout>(R.id.textinput_password).error).isNull()
    verify(activity.userServices).login(eq("username"), eq("password"), eq(false), callback.capture())
    callback.value.onDone(true)
    assertThat(activity).isFinishing
    assertThat(ShadowToast.getTextOfLatestToast()).isEqualTo(activity.getString(R.string.welcome, "username"))
    assertThat<Account>(AccountManager.get(activity).accounts).hasSize(1)
    assertThat(Preferences.getUsername(activity)).isEqualTo("username")
  }

  @Test
  fun testRegisterFailed() {
    activity.findViewById<EditText>(R.id.edittext_username).setText("username")
    activity.findViewById<EditText>(R.id.edittext_password).setText("password")
    activity.findViewById<View>(R.id.register_button).performClick()
    verify(activity.userServices).login(eq("username"), eq("password"), eq(true), callback.capture())
    callback.value.onDone(false)
    assertThat(activity).isNotFinishing
    assertThat(ShadowToast.getTextOfLatestToast()).isEqualTo(activity.getString(R.string.login_failed))
  }

  @Test
  fun testLoginError() {
    activity.findViewById<EditText>(R.id.edittext_username).setText("username")
    activity.findViewById<EditText>(R.id.edittext_password).setText("password")
    activity.findViewById<View>(R.id.login_button).performClick()
    verify(activity.userServices).login(eq("username"), eq("password"), eq(false), callback.capture())
    callback.value.onError(IOException())
    assertThat(activity).isNotFinishing
    assertThat(ShadowToast.getTextOfLatestToast()).isEqualTo(activity.getString(R.string.login_failed))
  }

  @Test
  fun testReLogin() {
    controller = Robolectric.buildActivity(TestLoginActivity::class.java)
    Preferences.setUsername(RuntimeEnvironment.application, "existing")
    activity = controller.create().get()
    assertThat(activity).hasTitle(R.string.re_enter_password)
    assertThat(activity.findViewById<View>(R.id.register_button)).isNotVisible
    assertThat(activity.findViewById<EditText>(R.id.edittext_username))
        .hasTextString("existing")
  }

  @Test
  fun testAddAccount() {
    Preferences.setUsername(RuntimeEnvironment.application, "existing")
    controller = Robolectric.buildActivity(TestLoginActivity::class.java,
        Intent().putExtra(LoginActivity.EXTRA_ADD_ACCOUNT, true))
    activity = controller.create().get()
    assertThat(activity.findViewById<View>(R.id.register_button)).isVisible
    assertThat(activity.findViewById<EditText>(R.id.edittext_username)).isEmpty
  }

  @After
  fun tearDown() {
    controller.destroy()
  }

  class TestLoginActivity : LoginActivity() {
    val userServices = mock(UserServices::class.java)!!

    override fun inject(any: Any) {
      mUserServices = userServices
      mAccountManager = AccountManager.get(RuntimeEnvironment.application)
    }
  }
}
